Add fetching of the last relaychain

janfabian opened the pull request:
and parachain blocks #81

<!--- Provide a general summary of your changes in the Title above -->
<!--- We are using semantinc pull request to make our lives easier -->
<!--- Please use prefixes for the title and use ! before : if breaking like below -->
<!--- build: ci: docs: feat: perf: refactor: fix: style: test: refactor!: -->

## Description
<!--- Describe your changes in detail -->

## Related Issue
<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue below using keyword like: Fixes: #0 -->

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have updated the documentation if necessary.
- [x] I have added tests to cover my changes, regression test if fixing an issue.
- [ ] This is a breaking change.

maht0rz commented on src/hooks/lastBlock/queries/useLastBlockQuery.tsx:
```diff
@@ -0,0 +1,11 @@
+  lastBlock?: LastBlock;
+}
+
+export const useLastBlockQuery = (options?: QueryHookOptions) =>
```
this query should be configured to use a cache-only fetch policy, we don't want it to try and hit a non existent local resolver

maht0rz commented on src/hooks/lastBlock/useRefetchWithNewBlock.tsx:
```diff
@@ -22,37 +26,48 @@ export const writeLastBlock = (
+    }
 
-    const lastBlockData = client?.cache.readQuery<GetLastBlockQueryResponse>({
+    if (!validationData) {
```
i'd put `chainLastBlock and validationData` together

maht0rz replied:
also `client`

maht0rz reviewed (changes requested):
Let's break it down, first we have our task #81 which should be represented in this PR. It should only rewrite how the `LastBlock` entity data is obtained. I'd suggest just implementing a single hook `useLastBlock()` or similar that would aggregate both parachain & relaychain block fetching. 

This means that we can break down the parachain block subscription & relaychain block fetching (per parachain block hash) into non-hook functions that should be easily testable.

Please try to figure out an appropriate separation of concerns, less hooks means easier testing in this case

maht0rz commented on src/hooks/lastBlock/useValidationData.tsx:
```diff
@@ -0,0 +1,44 @@
+    }
+
+    const getValidationData = async () => {
+      const rawValidationData =
```
here we should fetch the validation data for the blockhash received in the relaychain last block, as described in the issue (#81)

janfabian commented:
@maht0rz I implemented changes based on the last call with @dexterslabor

- fixed the multiple lastBlock query refetch
- refactored tests
 
please feel free to check it

maht0rz commented on src/hooks/lastBlock/lib/getValidationData.test.tsx:
```diff
@@ -0,0 +1,67 @@
+      parachainBlock
+    );
+
+    expect(jsonValidationData).toEqual({ mockJsonValidationData });
```
i'd prefer we'd also mock & check that the validation data is coming from the `.at` instance. Lets keep it as it is now, but down the line i'd rather test it like that

maht0rz commented on src/hooks/lastBlock/lib/getValidationData.test.tsx:
```diff
@@ -0,0 +1,67 @@
+  });
+
+  it(`doesn't get validation data from parachain block head`, async () => {
+    (mockApiInstance.createType as jest.Mock).mockReturnValueOnce({
```
Alternative to this would be using `as unknown as jest.Mocked<ApiPromise>` at the time of the mock setup. Then you can mock the return value as follows:

```
mockApiInstance.createType.mockReturnValueOnce({
      isSome: false,
    } as any as Codec);
```

maht0rz replied:
nvm, i'll push those changes myself and we can forget about it.

maht0rz commented on src/hooks/lastBlock/lib/getValidationData.tsx:
```diff
@@ -0,0 +1,50 @@
+  return apiInstanceSpecific;
+};
+
+// TODO: add test and extract to some common utils
```
tests are missing for certain helpers, but we can let it fly this time

mckrava commented:
Storybook testing artifacts:
- [storybook-ci-tests-results](https://github.com/galacticcouncil/Basilisk-ui/suites/5346415887/artifacts/167188763);
Or check GH Action `Storybook testing CI flow` for more logs.

mckrava commented:
Storybook testing artifacts:
- [storybook-ci-tests-results](https://github.com/galacticcouncil/Basilisk-ui/suites/5357619805/artifacts/167812562);
Or check GH Action `Storybook testing CI flow` for more logs.

mckrava commented:
Storybook testing artifacts:
- [storybook-ci-tests-results](https://github.com/galacticcouncil/Basilisk-ui/suites/5358040190/artifacts/167837887);
Or check GH Action `Storybook testing CI flow` for more logs.

maht0rz commented:
We're gonna use the old implementation for the time being.
